File: Java/Medium/3212 Count Submatrices With Equal Frequency of X and Y.java
class Solution {
    public int numberOfSubmatrices(char[][] grid) {
        int C = grid[0].length;
        int[] diff = new int[C];
        boolean[] hasX = new boolean[C];
        int count = 0;
        
        // Iterate directly over rows to cache the 1D array reference
        for (char[] row : grid) {
            int rowDiff = 0;
            boolean rowHasX = false;
            
            for (int j = 0; j < C; j++) {
                char val = row[j];
                
                if (val == 'X') {
                    rowDiff++;
                    rowHasX = true;
                } else if (val == 'Y') {
                    rowDiff--;
                }
                
                // Update vertical prefix state
                diff[j] += rowDiff;
                
                // Once a column sees an 'X', it remains true
                if (rowHasX) {
                    hasX[j] = true;
                }
                
                // Check validity
                if (diff[j] == 0 && hasX[j]) {
                    count++;
                }
            }
        }
        return count;
    }
}

File: Python/Medium/3212 Count Submatrices With Equal Frequency of X and Y.py
class Solution:
    def numberOfSubmatrices(self, grid: List[List[str]]) -> int:
        C = len(grid[0])
        diff = [0] * C
        has_x = [False] * C
        ans = 0
        
        for row in grid:
            row_diff = 0
            row_has_x = False
            
            for c, val in enumerate(row):
                if val == 'X':
                    row_diff += 1
                    row_has_x = True
                elif val == 'Y':
                    row_diff -= 1
                
                diff[c] += row_diff
                if row_has_x:
                    has_x[c] = True
                
                if diff[c] == 0 and has_x[c]:
                    ans += 1
                    
        return ans
